File: java_versions/v1_serial.java
package java_versions;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.*;

public class v1_serial {

    public static void main(String[] args) {
        String caminhoArquivo = "corpus_grande.txt";
        String arquivoSaida = "resultados/v1_resultado_tfidf_geral.csv";
        
        Map<String, Integer> documentosComTermo = new HashMap<>();
        Map<String, Double> tfIdfAcumuladoGlobal = new HashMap<>();
        long totalDocumentos = 0;

        // Passo 1: Calculando frequencia documental (IDF)...
        try (BufferedReader br = new BufferedReader(new FileReader(caminhoArquivo))) {
            String linha;
            while ((linha = br.readLine()) != null) {
                totalDocumentos++;
                String[] palavras = linha.toLowerCase().split("\\s+");
                
                Set<String> termosUnicosNoDoc = new HashSet<>(Arrays.asList(palavras));
                for (String termo : termosUnicosNoDoc) {
                    if (!termo.isEmpty()) {
                        documentosComTermo.put(termo, documentosComTermo.getOrDefault(termo, 0) + 1);
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Passo 2: Calculando TF-IDF por linha e acumulando...
        try (BufferedReader br = new BufferedReader(new FileReader(caminhoArquivo))) {
            String linha;
            while ((linha = br.readLine()) != null) {
                String[] palavras = linha.toLowerCase().split("\\s+");
                if (palavras.length == 0) continue;

                Map<String, Integer> frequenciaTermo = new HashMap<>();
                for (String p : palavras) {
                    if (!p.isEmpty()) {
                        frequenciaTermo.put(p, frequenciaTermo.getOrDefault(p, 0) + 1);
                    }
                }

                for (String termo : frequenciaTermo.keySet()) {
                    double tf = (double) frequenciaTermo.get(termo) / palavras.length;
                    double idf = Math.log((double) totalDocumentos / documentosComTermo.get(termo));
                    double tfidf = tf * idf;

                    // Acumula o valor para posterior média
                    tfIdfAcumuladoGlobal.put(termo, tfIdfAcumuladoGlobal.getOrDefault(termo, 0.0) + tfidf);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        // Passo 3: Calculando médias e salvando resultados no arquivo de saida...
        try (BufferedWriter bw = new BufferedWriter(new FileWriter(arquivoSaida))) {
            bw.write("Palavra;TF-IDF_Medio\n");

            // Ordenando por valor (opcional, mas ajuda a ver as mais importantes no topo do arquivo)
            List<Map.Entry<String, Double>> listaOrdenada = new ArrayList<>(tfIdfAcumuladoGlobal.entrySet());
            listaOrdenada.sort((a, b) -> b.getValue().compareTo(a.getValue()));

            for (Map.Entry<String, Double> entrada : listaOrdenada) {
                // MÉDIA: Divide o acumulado pelo total de documentos do corpus
                double mediatfidf = entrada.getValue() / totalDocumentos;
                
                bw.write(String.format("%s;%.10f\n", entrada.getKey(), mediatfidf));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: nim/v1_serial.nim
import std/[strutils, tables, math, algorithm, sets, times, os, strformat]

proc main() =
  let caminhoArquivo = "corpus_grande.txt"
  let arquivoSaida = "resultado_tfidf_geral.csv"

  var documentosComTermo = initCountTable[string]()
  var tfIdfAcumuladoGlobal = initTable[string, float64]()
  var totalDocumentos: int64 = 0

  # --- Passo 1: Frequência Documental (IDF) ---
  echo "Passo 1: Calculando frequencia documental (IDF)..."
  if not fileExists(caminhoArquivo):
    echo "Erro: Arquivo não encontrado."
    return

  for linha in lines(caminhoArquivo):
    totalDocumentos += 1
    let palavras = linha.toLowerAscii().splitWhitespace()
    
    # Set para termos únicos na linha (documento)
    var termosUnicosNoDoc = initHashSet[string]()
    for p in palavras:
      if p.len > 0:
        termosUnicosNoDoc.incl(p)
    
    for termo in termosUnicosNoDoc:
      documentosComTermo.inc(termo)

  # --- Passo 2: Calculando TF-IDF e acumulando ---
  echo "Passo 2: Calculando TF-IDF por linha e acumulando..."
  for linha in lines(caminhoArquivo):
    let palavras = linha.toLowerAscii().splitWhitespace()
    if palavras.len == 0: continue

    var frequenciaTermo = initCountTable[string]()
    for p in palavras:
      if p.len > 0:
        frequenciaTermo.inc(p)

    for termo, count in frequenciaTermo.pairs:
      let tf = count.float64 / palavras.len.float64
      let idf = ln(totalDocumentos.float64 / documentosComTermo[termo].float64)
      let tfidf = tf * idf
      
      tfIdfAcumuladoGlobal[termo] = tfIdfAcumuladoGlobal.getOrDefault(termo, 0.0) + tfidf

  # --- Passo 3: Médias e Exportação ---
  echo "Passo 3: Calculando médias e salvando resultados em ", arquivoSaida, "..."
  
  var f: File
  if open(f, arquivoSaida, fmWrite):
    f.writeLine("Palavra;TF-IDF_Medio")

    # Criar lista para ordenação (decrescente por valor)
    var listaOrdenada: seq[(string, float64)] = @[]
    for k, v in tfIdfAcumuladoGlobal.pairs:
      listaOrdenada.add((k, v))
    
    listaOrdenada.sort(proc (x, y: (string, float64)): int =
      cmp(y[1], x[1])
    )

    for item in listaOrdenada:
      let mediaTfIdf = item[1] / totalDocumentos.float64
      # Formatando com 10 casas decimais como no Java
      f.writeLine(&"{item[0]};{mediaTfIdf:.10f}")
    
    f.close()

  echo "=== RELATÓRIO FINAL ==="
  echo "Total de Linhas: ", totalDocumentos
  echo "Total de Palavras Únicas: ", tfIdfAcumuladoGlobal.len
  echo "O arquivo CSV foi gerado com sucesso."

main()